### ISSUE TITLE
Look into improving the 64Drive's USB protocol

### ISSUE THREAD

**Issue by buu342** [buu342/N64-UNFLoader#51]
1. The arming process sometimes causes stuff to go missing. This is because of firmware bug documented in the wiki. Sadly, this is up to Marshall to fix...
2. ~~You **should** be able to write data larger than 512 bytes at once without problems, **as long as it's 512 byte aligned**. The hardware spec is wrong, and the wiki needs to be updated.~~

**Comment by buu342** [comment]
Fixed Number 2 in 4d2a6e1

**Comment by buu342** [comment]
A possible fix for 1 would be to continually send the data over and over until the N64 responds with an ACK, which forces the cartridge to stall until it receives data again. I would prefer this solution to the current one of data potentially "missing"

**Comment by buu342** [comment]
Looks like this entire time it was due to me missing out a step: I had to wait for the arm mask to be 0x02.

```
while ((usb_io_read(D64_REG_USBCOMSTAT) & D64_CUI_ARM_MASK) != D64_CUI_ARM_UNARMED_DATA);
```

This was the key. Sorry Marshall, this one's on me!

**Comment by buu342** [comment]
Fixed in [c4178b9](https://github.com/buu342/N64-UNFLoader/commit/c4178b99629ad3e6e7ea8ba387939c2ca04eefc0).
